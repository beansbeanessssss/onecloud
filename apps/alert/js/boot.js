define({

	 /**
 	 * allow to setup the application
 	 *
 	 * @return promise
	 * @param self will feed back 'this' object
 	 */

	setup: () => {
		var p = new Promise((r, d) => {
			requirejs([OC.appJS('alert', 'alert.bundle')], () => {
				r({
					id: 'alert',
					start : true
				});
			});
		});
		return p;
	},

	boot: () => { return false }
});
