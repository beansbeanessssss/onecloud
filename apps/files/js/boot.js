define({

	 	/**
		 * allow to setup the application
		 *
		 * @return promise
		 * @param self will feed back 'this' object
		 */

	 	setup: () => {
	 		var my = self.info;
	 		var p = new Promise((resolve, defer) => {
	 			OC.registerNavItem('files', {
	 				name: 'Files',
	 				iconMaterial: 'folder',
	 				route: '/'
	 			});
				resolve( { id : 'files' } );
	 		});
	 		return p;
	 	},


		/**
		 * Start the application by mounting it to the respective DOM element
		 *
		 * @return promise
		 */

		boot: (container) => {
			var p = new Promise((resolve, defer) => {
				requirejs([OC.appJS('files', 'files.bundle')], (app) => {
					app.$mount(container);
					app.$once('mounted', resolve( { start: true } ) );
				});
			});
			return p;
		}
});
