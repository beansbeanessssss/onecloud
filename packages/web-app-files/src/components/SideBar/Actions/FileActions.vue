<template>
  <oc-list id="oc-files-actions-sidebar" class="oc-mt-s">
    <action-menu-item
      v-for="(action, index) in actions"
      :key="`action-${index}`"
      :action="action"
      :items="resources"
      class="oc-py-xs"
    />
  </oc-list>
</template>

<script>
import { mapGetters } from 'vuex'
import ActionMenuItem from '../../ActionMenuItem.vue'

import FileActions from '../../../mixins/fileActions'
import { useIsFilesAppActive } from '../../../composables/useIsFilesAppActive'

export default {
  name: 'FileActions',
  components: { ActionMenuItem },
  mixins: [FileActions],
  setup() {
    return {
      isFilesAppActive: useIsFilesAppActive()
    }
  },
  title: ($gettext) => {
    return $gettext('Actions')
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile']),

    resources() {
      return [this.highlightedFile]
    },

    actions() {
      return this.$_fileActions_getAllAvailableActions(this.resources)
    }
  }
}
</script>

<style lang="scss">
#oc-files-actions-sidebar {
  > li a,
  > li a:hover {
    color: var(--oc-color-swatch-passive-default);
    display: inline-flex;
    gap: 10px;
    vertical-align: top;
  }
}
</style>
