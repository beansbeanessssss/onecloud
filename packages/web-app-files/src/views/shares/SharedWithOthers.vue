<template>
  <div class="oc-flex">
    <files-view-wrapper>
      <app-bar :side-bar-open="sideBarOpen">
        <template #navigation>
          <SharesNavigation />
        </template>
      </app-bar>
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <div v-if="shareTypes.length > 1" class="oc-flex oc-m-m">
          <div class="oc-mr-m oc-flex oc-flex-middle">
            <oc-icon name="filter-2" class="oc-mr-xs" />
            <span v-text="$gettext('Filter:')" />
          </div>
          <item-filter
            :allow-multiple="true"
            :filter-label="$gettext('Share Type')"
            :filterable-attributes="['label']"
            :items="shareTypes"
            :option-filter-label="$gettext('Filter share types')"
            :show-option-filter="true"
            id-attribute="key"
            class="share-type-filter oc-mx-s"
            display-name-attribute="label"
            filter-name="shareType"
          >
            <template #item="{ item }">
              <span class="oc-ml-s" v-text="item.label" />
            </template>
          </item-filter>
        </div>
        <no-content-message
          v-if="isEmpty"
          id="files-shared-with-others-empty"
          class="files-empty"
          icon="reply"
        >
          <template #message>
            <span v-translate> You have not shared any resources with other people. </span>
          </template>
        </no-content-message>
        <resource-table
          v-else
          id="files-shared-with-others-table"
          v-model:selectedIds="selectedResourcesIds"
          class="files-table"
          :class="{ 'files-table-squashed': sideBarOpen }"
          :fields-displayed="['name', 'sharedWith', 'sdate']"
          :are-thumbnails-displayed="displayThumbnails"
          :are-paths-displayed="true"
          :resources="filteredItems"
          :header-position="fileListHeaderY"
          :sort-by="sortBy"
          :sort-dir="sortDir"
          :grouping-settings="groupingSettings"
          @file-click="triggerDefaultAction"
          @row-mounted="rowMounted"
          @sort="handleSort"
        >
          <template #contextMenu="{ resource }">
            <context-actions
              v-if="isResourceInSelection(resource)"
              :action-options="{ space: getMatchingSpace(resource), resources: selectedResources }"
            />
          </template>
          <template #footer>
            <pagination :pages="paginationPages" :current-page="paginationPage" />
            <list-info
              v-if="filteredItems.length > 0"
              class="oc-width-1-1 oc-my-s"
              :files="totalFilesCount.files"
              :folders="totalFilesCount.folders"
            />
          </template>
        </resource-table>
      </template>
    </files-view-wrapper>
    <side-bar
      :open="sideBarOpen"
      :active-panel="sideBarActivePanel"
      :space="selectedResourceSpace"
    />
  </div>
</template>

<script lang="ts">
import { mapGetters, mapState, mapActions } from 'vuex'

import { queryItemAsString, useFileActions, useRouteQuery } from '@ownclouders/web-pkg'
import { VisibilityObserver, ItemFilter } from '@ownclouders/web-pkg'
import { ImageDimension, ImageType } from '@ownclouders/web-pkg'
import { debounce, find, uniq } from 'lodash-es'

import { ResourceTable } from '@ownclouders/web-pkg'
import { AppLoadingSpinner } from '@ownclouders/web-pkg'
import { NoContentMessage } from '@ownclouders/web-pkg'
import { AppBar } from '@ownclouders/web-pkg'
import ListInfo from '../../components/FilesList/ListInfo.vue'
import { Pagination } from '@ownclouders/web-pkg'
import { ContextActions } from '@ownclouders/web-pkg'
import SideBar from '../../components/SideBar/SideBar.vue'
import FilesViewWrapper from '../../components/FilesViewWrapper.vue'

import { useResourcesViewDefaults } from '../../composables'
import { defineComponent, computed, unref } from 'vue'
import { Resource } from '@ownclouders/web-client'
import { useGroupingSettings } from '@ownclouders/web-pkg'
import { useGetMatchingSpace, useMutationSubscription } from '@ownclouders/web-pkg'
import SharesNavigation from 'web-app-files/src/components/AppBar/SharesNavigation.vue'
import { ShareTypes } from '@ownclouders/web-client/src/helpers'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    SharesNavigation,
    FilesViewWrapper,
    AppBar,
    ResourceTable,
    AppLoadingSpinner,
    NoContentMessage,
    ListInfo,
    Pagination,
    ContextActions,
    SideBar,
    ItemFilter
  },

  setup() {
    const { getMatchingSpace } = useGetMatchingSpace()

    const resourcesViewDefaults = useResourcesViewDefaults<Resource, any, any[]>()
    const { sortBy, sortDir, loadResourcesTask, selectedResourcesIds, paginatedResources } =
      resourcesViewDefaults

    const shareTypes = computed(() => {
      const uniqueShareTypes = uniq(unref(paginatedResources).map((i) => i.share?.shareType))
      return ShareTypes.getByValues(uniqueShareTypes)
    })
    const selectedShareTypesQuery = useRouteQuery('q_shareType')
    const filteredItems = computed(() => {
      const selectedShareTypes = queryItemAsString(unref(selectedShareTypesQuery))?.split('+')
      if (!selectedShareTypes || selectedShareTypes.length === 0) {
        return unref(paginatedResources)
      }
      return unref(paginatedResources).filter((item) => {
        return selectedShareTypes.map((t) => ShareTypes[t].value).includes(item.share.shareType)
      })
    })

    useMutationSubscription(['Files/UPDATE_RESOURCE_FIELD'], async (mutation) => {
      if (mutation.payload.field === 'shareTypes') {
        if (selectedResourcesIds.value.length !== 1) return
        const id = selectedResourcesIds.value[0]

        const match = find(paginatedResources.value, { id })
        if (!match) return

        await loadResourcesTask.perform()

        const matchedNewResource = find(paginatedResources.value, { fileId: match.fileId })
        if (!matchedNewResource) return

        selectedResourcesIds.value = [matchedNewResource.id]
      }
    })

    return {
      ...useFileActions(),
      ...resourcesViewDefaults,
      filteredItems,
      shareTypes,
      getMatchingSpace,

      // CERN
      ...useGroupingSettings({ sortBy, sortDir })
    }
  },

  computed: {
    ...mapState(['app']),
    ...mapState('Files', ['files']),
    ...mapGetters('Files', ['totalFilesCount']),
    ...mapGetters(['configuration', 'user']),

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    displayThumbnails() {
      return !this.configuration?.options?.disablePreviews
    }
  },

  async created() {
    await this.loadResourcesTask.perform()
    this.scrollToResourceFromRoute(this.paginatedResources, 'files-app-bar')
  },

  beforeUnmount() {
    visibilityObserver.disconnect()
  },

  methods: {
    ...mapActions('Files', ['loadPreview', 'loadAvatars']),

    rowMounted(resource, component) {
      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadAvatars({ resource, clientService: this.$clientService })

        if (!this.displayThumbnails) {
          return
        }

        this.loadPreview({
          previewService: this.$previewService,
          space: this.getMatchingSpace(resource),
          resource,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
