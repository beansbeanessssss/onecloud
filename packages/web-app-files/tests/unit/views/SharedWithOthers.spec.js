import { mount, RouterLinkStub } from '@vue/test-utils'
import { getStore, localVue, createFile, stubs } from './views.setup.js'
import FileActions from '@files/src/mixins/fileActions.js'
import SharedWithOthers from '@files/src/views/SharedWithOthers.vue'
import MixinMountSideBar from '@files/src/mixins/sidebar/mountSideBar.js'

const listLoaderStub = 'list-loader-stub'
const noContentStub = 'no-content-message-stub'
const filesTableStub = 'oc-table-files-stub'
const contextActionsStub = 'context-actions-stub'
const listInfoStub = 'list-info-stub'

const selectors = {
  pagination: '.files-pagination'
}

describe('SharedWithOthers', () => {
  it('should show list loader when loading is set as true', () => {
    const wrapper = getMountedWrapper({ loading: true })

    expect(wrapper.find(listLoaderStub).exists()).toBeTruthy()
    expect(wrapper.find(noContentStub).exists()).toBeFalsy()
    expect(wrapper.find(filesTableStub).exists()).toBeFalsy()
  })
  describe('when loading is set as false', () => {
    it('should show the no content message component if active files length is less than one', () => {
      const store = createStore({ activeFiles: [] })
      const wrapper = getMountedWrapper({ store, loading: false })

      expect(wrapper.find(listLoaderStub).exists()).toBeFalsy()
      expect(wrapper.find(filesTableStub).exists()).toBeFalsy()
      expect(wrapper.find(noContentStub).exists()).toBeTruthy()
      expect(wrapper.find(noContentStub)).toMatchSnapshot()
    })
    describe('when length of active files is greater than zero', () => {
      const activeFiles = [createFile({ id: '1234' }), createFile({ id: '5896' })]
      const store = createStore({ activeFiles: activeFiles })
      const wrapper = getMountedWrapper({ store, loading: false })

      it('should not show the no content message component', () => {
        expect(wrapper.find(noContentStub).exists()).toBeFalsy()
      })

      it('should load files table with correct props', () => {
        const filesTable = wrapper.find(filesTableStub)

        expect(filesTable.exists()).toBeTruthy()
        expect(filesTable.props()).toMatchObject({
          resources: activeFiles,
          arePathsDisplayed: false,
          areThumbnailsDisplayed: !store.getters.configuration.options.disablePreviews,
          areResourcesClickable: true,
          isSelectable: true,
          disabled: null,
          dragDrop: false,
          selection: [],
          hasActions: true,
          targetRoute: { name: 'files-personal' }
        })
      })

      it('should show context menu for each resource', () => {
        stubs['oc-table-files'] = false
        stubs['context-actions'] = true
        const wrapper = getMountedWrapper({ store, loading: false })
        const contextActions = wrapper.findAll(contextActionsStub)

        expect(contextActions.length).toBe(activeFiles.length)
        activeFiles.forEach((activeFile, index) => {
          expect(contextActions.at(index).props()).toMatchObject({
            item: activeFile
          })
        })
      })

      it('should show list info component for oc files table', () => {
        stubs['oc-table-files'] = false
        stubs['list-info'] = true

        const store = createStore({
          activeFiles: activeFiles,
          totalFilesCount: {
            files: 22,
            folders: 56
          }
        })

        const wrapper = getMountedWrapper({ store, loading: false })
        const listInfo = wrapper.find(listInfoStub)

        expect(listInfo.exists()).toBeTruthy()
        expect(listInfo.props()).toMatchObject({
          files: store.getters['Files/totalFilesCount'].files,
          folders: store.getters['Files/totalFilesCount'].folders,
          size: null
        })
      })

      describe('pagination component in oc files table', () => {
        it('should be visible if pages is greater than one', () => {
          stubs['oc-table-files'] = false
          stubs.pagination = false
          stubs.RouterLink = RouterLinkStub
          const store = getStore({ activeFiles: activeFiles, pages: 2, currentPage: 1 })
          const wrapper = getMountedWrapper({ store, loading: false })

          const pagination = wrapper.find(selectors.pagination)

          expect(pagination.exists()).toBeTruthy()
          stubs['oc-table-files'] = true
        })

        it('should not be visible if pages is less than one', () => {
          stubs['oc-table-files'] = false
          stubs.pagination = false
          stubs.RouterLink = RouterLinkStub
          const store = getStore({ activeFiles: activeFiles, pages: 1, currentPage: 1 })
          const wrapper = getMountedWrapper({ store, loading: false })

          const pagination = wrapper.find(selectors.pagination)

          expect(pagination.exists()).toBeFalsy()
          stubs['oc-table-files'] = true
        })
      })

      it('should call "$_mountSideBar_showDefaultPanel" method if "showDetails" event is emitted from oc files table component', () => {
        const spyMountSideBarShowDefaultPanel = jest.spyOn(
          MixinMountSideBar.methods,
          '$_mountSideBar_showDefaultPanel'
        )
        const wrapper = getMountedWrapper({ store, loading: false })
        console.log(wrapper.html())
        const filesTable = wrapper.find(filesTableStub)
        expect(spyMountSideBarShowDefaultPanel).toHaveBeenCalledTimes(0)

        filesTable.vm.$emit('showDetails')

        expect(spyMountSideBarShowDefaultPanel).toHaveBeenCalledTimes(1)
      })

      it('should call "$_fileActions_triggerDefaultAction" method if "fileClick" event is emitted from oc files table component', () => {
        const spyFileActionsTriggerDefaultAction = jest
          .spyOn(FileActions.methods, '$_fileActions_triggerDefaultAction')
          .mockImplementation()
        const wrapper = getMountedWrapper({ store, loading: false })
        const filesTable = wrapper.find(filesTableStub)
        expect(spyFileActionsTriggerDefaultAction).toHaveBeenCalledTimes(0)

        filesTable.vm.$emit('fileClick')

        expect(spyFileActionsTriggerDefaultAction).toHaveBeenCalledTimes(1)
      })

      it('should call "rowMounted" method if "rowMounted" event is emitted from oc files table component', () => {
        const spyRowMounted = jest
          .spyOn(SharedWithOthers.methods, 'rowMounted')
          .mockImplementation()
        const wrapper = getMountedWrapper({ store, loading: false })
        const filesTable = wrapper.find(filesTableStub)
        expect(spyRowMounted).toHaveBeenCalledTimes(0)

        filesTable.vm.$emit('rowMounted')

        expect(spyRowMounted).toHaveBeenCalledTimes(1)
      })
    })
  })

  function createStore({
    activeFiles = [],
    sidebarClosed = true,
    disablePreviews = false,
    totalFilesCount
  } = {}) {
    return getStore({ activeFiles, sidebarClosed, disablePreviews, totalFilesCount })
  }

  function getMountOptions({
    $route = { params: { page: 1 } },
    store = createStore(),
    loading = true
  }) {
    return {
      localVue,
      store,
      stubs,
      mocks: {
        $route
      },
      data: () => ({
        loading: loading
      })
    }
  }

  function getMountedWrapper({ store, loading } = {}) {
    const component = { ...SharedWithOthers, created: jest.fn(), mounted: jest.fn() }
    return mount(component, getMountOptions({ store, loading }))
  }
})
