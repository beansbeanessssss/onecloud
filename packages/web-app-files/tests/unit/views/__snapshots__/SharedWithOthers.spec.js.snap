// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedWithOthers when loading is set as false should show the no content message component if active files length is less than one 1`] = `<no-content-message-stub icon="group" id="files-shared-with-others-empty" class="files-empty"></no-content-message-stub>`;
